<launch>
    <arg name="launch_prefix" default="" />
    <arg name="node_namespace" default="apriltag_ros_continuous_node" />
    <arg name="camera_name" default="/xtion/rgb" />
    <arg name="image_topic" default="image_raw" />
    <arg name="tag_location_file" default="tag_world.yaml"/>

    <arg name="localization_method" default="weighted_average"/>
    <arg name="buffer_size" default="10"/>
    <arg name="max_time_diff" default="0.2"/>
    <arg name="max_error" default="0.2"/>
    <arg name="filter_mode" default="mean"/>

    <arg name="save_location" default="$(find apriltag_localization)/turn_tests/mean/data"/>
    <arg name="save_name" default="test_buf_er_tdiff"/>
    <arg name="save_format" default=".csv"/>


    <rosparam command="load" file="$(find apriltag_ros)/config/settings.yaml" ns="$(arg node_namespace)" />
    <rosparam command="load" file="$(find apriltag_ros)/config/tags.yaml" ns="$(arg node_namespace)" />
    
    <rosparam command="load" file="$(find apriltag_localization)/config/$(arg tag_location_file)" ns="apriltag_localization" />
    
    <!-- General args -->
    <arg name="rviz" default="false"/>
    <arg name="gzclient" default="false"/>
    <arg name="gzpose" default="-x -1.0 -y 2.0 -z 0.0 -R 0.0 -P 0.0 -Y -1.5707" />
    <arg name="world" default="turn_test"/>
    <arg name="map" default="$(find retail_store_simulation)/maps/full_demo"/>
    <arg name="use_aruco" default="false"/>
    <arg name="use_moveit_camera" default="true"/>
    <arg name="robot" default="tiago"/>

    <!-- Launch sim with robot -->
    <include file="$(find retail_store_simulation)/launch/tiago.launch" if="$(eval arg('robot') == 'tiago')" pass_all_args="true" />
    <include file="$(find retail_store_simulation)/launch/albert.launch" if="$(eval arg('robot') == 'albert')" pass_all_args="true" />
 
    <!-- Launch grocery store utils -->
    <group if="$(eval arg('use_aruco'))">
        <node pkg="grocery_store_utils" name="collision_object_server" type="collision_object_server.py" output="screen" /> 
        <node pkg="grocery_store_utils" name="grasp_pose_server" type="grasp_pose_server.py" output="screen" /> 
    </group>


    <node pkg="apriltag_ros" type="apriltag_ros_continuous_node" name="$(arg node_namespace)" clear_params="true" output="screen" launch-prefix="$(arg launch_prefix)" >
        <remap from="image_rect" to="$(arg camera_name)/$(arg image_topic)" />
        <remap from="camera_info" to="$(arg camera_name)/camera_info" />
        <param name="publish_tag_detections_image" type="bool" value="true" /> 
    </node>

    <node pkg="apriltag_localization" name="tag_localization_node" type="simulation_detection_node.py" output="screen">
        <param name="localization_method" type="string" value="$(arg localization_method)" />
        <param name="buffer_size" type="int" value="$(arg buffer_size)" />
        <param name="max_time_diff" type="double" value="$(arg max_time_diff)" />
        <param name="max_error" type="double" value="$(arg max_error)" />
        <param name="filter_mode" type="string" value="$(arg filter_mode)" />
    </node>

    <node pkg="apriltag_localization" name="simulation_turn_test" type="simulation_turn_test.py" output="screen">
        <param name="save_location" type="string" value="$(arg save_location)"/>
        <param name="save_name" type="string" value="$(arg save_name)"/>
        <param name="save_format" type="string" value="$(arg save_format)"/>
    </node>
</launch>